src/components/postgresConnect.ts

```diff
@@ -1,45 +1,210 @@
+    payload: ActionQueueGetPayload = { status: 'Scheduled' }
+  ): Promise<ActionQueue[]> => {
+    try {
+      const result = await this.query(
+        'SELECT id, action_code, parameters, execution_time FROM action_queue WHERE status = $1 ORDER BY execution_time',
+        Object.values(payload)
+      )
+      return result.rows as ActionQueue[]
+    } catch (err) {
+      return err.stack
+    }
+  }
 
-pgClient.on('notification', (msg: DatabaseRecord) => {
-  switch (msg.channel) {
-    case 'trigger_notifications':
-      processTrigger(pgClient, JSON.parse(msg.payload))
-      break
-    case 'action_notifications':
-      executeAction(pgClient, JSON.parse(msg.payload), actionLibrary)
-      break
+  public addFile = async (payload: FilePayload): Promise<Pick<File, 'id'>> => {
+    try {
+      const result = await this.query(
+        `INSERT INTO file (${Object.keys(payload)}) VALUES () RETURNING id`,
+        Object.values(payload)
+      )
+      return result.rows[0] as Pick<File, 'id'>
+    } catch (err) {
+      return err.stack
+    }
   }
-  // console.log(msg.payload);
-})
 
-pgClient.query('LISTEN trigger_notifications')
-pgClient.query('LISTEN action_notifications')
+  public getFile = async (payload: FileGetPayload): Promise<File> => {
+    try {
+      const result = await this.query(
+        'SELECT path, original_filename FROM file WHERE id = $1',
+        Object.values(payload)
+      )
+      return result.rows[0] as File
+    } catch (err) {
+      return err.stack
+    }
+  }
+
+  public addActionPlugin = async (plugin: ActionPluginPayload) => {
+    try {
+      await this.query(
+        `INSERT INTO action_plugin (${Object.keys(plugin)}) VALUES (${this.getValuesPlaceholders(
+          plugin
+        )})`,
+        Object.values(plugin)
+      )
+    } catch (err) {
+      console.log(err.stack)
+    }
+  }
+
+  public deleteActionPlugin = async (payload: ActionPluginDeletePayload) => {
+    try {
+      await this.query('DELETE FROM action_plugin WHERE code = $1', Object.values(payload))
+    } catch (err) {
+      console.log(err.stack)
+    }
+  }
+
+  public getActionPlugins = async (): Promise<ActionPlugin[]> => {
+    try {
+      const result = await this.query('SELECT * FROM action_plugin', [])
+      return result.rows as ActionPlugin[]
+    } catch (err) {
+      return err.stack
+    }
+  }
+
+  public getActionPluginsByTemplate = async (
+    tableName: string,
+    payload: ActionInTemplateGetPayload
+  ): Promise<ActionInTemplate[]> => {
+    try {
+      const result = await this.query(
+        `SELECT action_plugin.code, action_plugin.path, action_plugin.name, trigger, condition, parameter_queries 
+          FROM template 
+         JOIN template_action ON template.id = template_action.template_id 
+         JOIN action_plugin ON template_action.action_code = action_plugin.code 
+         WHERE template_id = (SELECT template_id FROM ${tableName} WHERE id = $1) AND trigger = $2`,
```

Comment:
This `${tableName}` substitution is the one place where the built-in parameter parser won't work. I had this problem too, but I now see that this is expected behaviour:

> PostgreSQL does not support parameters for identifiers. If you need to have dynamic database, schema, table, or column names (e.g. in DDL statements) use pg-format package for handling escaping these values to ensure you do not have SQL injection!

It's probably not worth installing `pg-format` for just this one thing (and it's a read query, so not particularly risky), but if you can't make the earlier substitutions work with proper parameterized queries, I'd suggest that it probably is worth it: https://www.npmjs.com/package/pg-format

Reply:
Hmm... Yes, I spent about 1 hour trying to understand this one the other day. But I still haven't seen how the `pg-format` could help with passing a dynamic set of { key, values } to the query. I'll see if I find something about it on the weekend.